File: LeetCode/0780. Reaching Points/solution.cpp
/**
 * 195 / 195 test cases passed.
 * Runtime: 4 ms
 * Memory Usage: 5.8 MB 
 */
class Solution {
public:
    bool reachingPoints(int sx, int sy, int tx, int ty) {
        while (tx >= sx && ty >= sy) {
            if (tx == sx && ty == sy) return true;
            if (tx > ty) {
                tx -= max((tx - sx) / ty, 1) * ty;
            } else {
                ty -= max((ty - sy) / tx, 1) * tx;
            }
        }
        return false;
    }
};

File: LeetCode/0780. Reaching Points/solution.rs
/**
 * 195 / 195 test cases passed.
 * Runtime: 0 ms
 * Memory Usage: 2 MB 
 */
impl Solution {
    pub fn reaching_points(sx: i32, sy: i32, tx: i32, ty: i32) -> bool {
        let (mut tx, mut ty) = (tx, ty);
        while tx >= sx && ty >= sy {
            if tx == sx && ty == sy {
                return true;
            }
            if tx > ty {
                if tx - sx < ty {
                    break;
                }
                tx -= ((tx - sx) / ty) * ty;
            } else {
                if ty - sy < tx {
                    break;
                }
                ty -= ((ty - sy) / tx) * tx;
            }
        }
        false
    }
}
